README.md
# openempi-demo

Some code making API calls to Open EMPI server.  Assumes Open EMPI
service to be available at `http://localhost:8090`.

Currently the only working "demo" code is the `demo-add-and-find`
function (see source).

## WARNING!!

__While this repository is made public, it really isn't meant for
public consumption.  The code here will not even compile unless you
have your maven repository set up properly with the correct version of
datomic library, for example.__

## License

Copyright © 2016 Bjond Inc.

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-12-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-12-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/openempi-demo/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/openempi-demo/compare/0.1.0...0.1.1


test/openempi_demo/core_test.clj
(ns openempi-demo.core-test
  (:require [clojure.test :refer :all]
            [openempi-demo.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject openempi-demo "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :repositories [["bjond-maven-private" { :url "http://nexus.bjondhealth.com/nexus/content/groups/private"}]
                 ["bjond-maven-public" { :url "http://nexus.bjondhealth.com/nexus/content/groups/public"}]]
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/data.json "0.2.6"]
                 [org.apache.httpcomponents/httpclient "4.3.5"]
                 [clj-http "2.3.0"]
                 [clj-time "0.12.2"]
                 [org.clojure/data.xml "0.0.8"]
                 [zprint "0.2.9"]
                 [com.datomic/datomic-pro "0.9.5390" :exclusions [joda-time]]]
  :main ^:skip-aot openempi-demo.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/openempi_demo/demo.clj
(ns openempi-demo.demo
  (:require [zprint.core :as zp]
            [clojure.data.json :as json]
            [clj-http.client :as http]
            [clojure.data.xml :as xml]
            [clj-time.core :as t]
            [openempi-demo.core :refer :all]))

(def p1 {"gender" {"genderCode" "M"}
         "familyName" "Cooper"
         "middleName" "Fenimore"
         "givenName" "James"
         "dateOfBirth" "2000-01-01"
         "personIdentifiers" {"identifier" "bjond_id_1"
                              "identifierDomain" {"namespaceIdentifier" "Bjond"}}})

(def p1-json (json/write-str p1))

(defn updated-person [person-map new-given-name]
  (-> person-map
      (assoc "givenName" new-given-name)
      (assoc "dateChanged" (.toString (t/now)))))

(def p2
  (xml/sexp-as-element
   ["person" {}
    ["gender" {} ["genderCode" {} "F"]]
    ["familyName" {} "Doe"]
    ["givenName" {} "Jane"]
    ["dateOfBirth" {} "2002-02-02"]]))

#_(def p3 "<?xml version=\"1.0\" encoding=\"UTF-8\"?><person>
  <dateChanged>2016-12-19T22:23:00.000Z</dateChanged>
  <dateCreated>2016-12-19T22:21:27.237Z</dateCreated>
  <dateOfBirth>2000-01-10T00:00:00Z</dateOfBirth>
  <familyName>Doe</familyName>
  <gender>
    <genderCd>1</genderCd>
    <genderCode>F</genderCode>
    <genderDescription>Female</genderDescription>
    <genderName>Female</genderName>
  </gender>
  <givenName>Johnny</givenName>
  <personIdentifiers>
    <dateCreated>2016-12-19T22:21:27.247Z</dateCreated>
    <identifier>7bb7fdf0-c639-11e6-ac4f-0242ac120003</identifier>
    <identifierDomain>
      <identifierDomainId>18</identifierDomainId>
      <identifierDomainName>OpenEMPI</identifierDomainName>
      <namespaceIdentifier>2.16.840.1.113883.4.357</namespaceIdentifier>
      <universalIdentifier>2.16.840.1.113883.4.357</universalIdentifier>
      <universalIdentifierTypeCode>hl7</universalIdentifierTypeCode>
    </identifierDomain>
  </personIdentifiers>
</person>")

(def key1 (xml/sexp-as-element ["person" {}
                                ["gender" {} ["genderCode" {} "M"]]
                                ["familyName" {} "Cooper"]
                                ["givenName" {} "James"]]))

(def key1-xml (xml/emit-str key1))
  
(defn demo-add-and-find [session-key]
  (try
    (let [response (add-person-json session-key p1-json)]
      (println "Added person:")
      (zp-body response))
    (println)
    (let [id (get-in p1 ["personIdentifiers" "identifier"])
          response (find-person-by-id session-key id)]
      (println "Found person:")
      (zp-body response))
    (println)
    (finally
      (let [response (delete-person session-key p1-json)]
        (println "Deleted person")
        (println :status (:status response))))))



    
  
  


src/openempi_demo/bjond.clj
(ns openempi-demo.bjond
  (:require [zprint.core :as zp]
            [clojure.data.json :as json]
            [clj-http.client :as http]))

(def url-base "http://localhost:8080")

(def p1 {"gender" {"genderCode" "M"}
         "familyName" "Cooper"
         "middleName" "Fenimore"
         "givenName" "James"
         "dateOfBirth" "2000-01-01"
         "personIdentifiers" {"identifier" "bjond_id_1"
                              "identifierDomain" {"namespaceIdentifier" "Bjond"}}})

(def p1-json (json/write-str p1))

(def key {"familyName" "Cooper"
          "givenName" "James"})

(def key-json (json/write-str key))

(defn add-person [json]
  (let [url (str url-base "/server-core/services/empiservice/add")]
    (http/post url {:body json})))

(defn import-person [json]
  (let [url (str url-base "/server-core/services/empiservice/import")]
    (http/post url {:body json})))

(defn find-person [id]
  (let [url (str url-base "/server-core/services/empiservice/find")]
    (http/get url {:query-params {"id" id
                                  "domainNamespaceID" "Bjond"}})))

(defn find-matches [json]
  (let [url (str url-base "/server-core/services/empiservice/findMatches")]
    (http/post url {:body json})))


src/openempi_demo/mbrainz.clj
(ns openempi-demo.mbrainz
  (:require [zprint.core :as zp]
            [clojure.data.json :as json]
            [clj-http.client :as http]
            [clojure.data.xml :as xml]
            [clj-time.core :as t]
            [datomic.api :as d]
            [openempi-demo.core :refer :all]))

(def db-uri "datomic:dev://localhost:4334/mbrainz-1968-1973")

(defn count-artists [db]
  (d/q '[:find (count ?e)
         :where [?e :artist/name _][?e :artist/type :artist.type/person]] db))

(defn count-artists-by-gender [db gender]
  (d/q [:find '(count ?e)
        :where ['?e :artist/gender gender]
        '[?e :artist/type :artist.type/person]] db))

(defn first-n-artist-names [db n]
  (take n (d/q '[:find ?n
                 :where [?e :artist/name ?n]
                 [?e :artist/type :artist.type/person]] db)))

(defn count-name-words [db]
  (let [names (d/q '[:find ?n
                     :where [?e :artist/name ?n]
                     [?e :artist/type :artist.type/person]] db)]
    (frequencies (map #(-> %1
                           first
                           (clojure.string/split #" ")
                           count) names))))

(defn filter-name-by-count [db n]
  (let [names (d/q '[:find ?n
                     :where [?e :artist/name ?n]
                     [?e :artist/type :artist.type/person]] db)]
    (filter #(-> %1
                 first
                 (clojure.string/split #" ")
                 count
                 (= n)) names)))

(defn two-named-gendered-with-date [db]
  (let [result (d/q '[:find ?e ?n ?g ?y ?m ?d
                      :where
                      [?e :artist/name ?n]
                      [?e :artist/type :artist.type/person]
                      [?e :artist/gender ?g]
                      [?e :artist/startYear ?y]
                      [?e :artist/startMonth ?m]
                      [?e :artist/startDay ?d]] db)]
    (filter #(-> %1
                 second
                 (clojure.string/split #" ")
                 count
                 (= 2)) result)))

(defn pull-by-name [db name]
  (d/q [:find '(pull ?e [*])
        :where ['?e :artist/name name]] db))

;; 17592186045591 :artist.gender/female
;; 17592186045654 :artist.gender/male

(defn musician-to-person [data]
  (let [[entity-id name gender-id year month day] data
        split-name (clojure.string/split name #" ")
        first-name (first split-name)
        last-name (second split-name)
        gender-code (if (= gender-id 17592186045591) "F" "M")
        date-string (format "%s-%02d-%02d" year month day)]
    #_(println :first-name first-name :last-name last-name :gender gender-code :date date-string)
    {"gender" {"genderCode" gender-code}
     "familyName" last-name
     "givenName" first-name
     "dateOfBirth" date-string
     "personIdentifiers" {"identifier" entity-id
                          "identifierDomain" {"namespaceIdentifier" "Bjond"}}}))

(defn musician-to-person [data domain-namespace]
  (let [[entity-id name gender-id year month day] data
        split-name (clojure.string/split name #" ")
        first-name (first split-name)
        last-name (second split-name)
        gender-code (if (= gender-id 17592186045591) "F" "M")
        date-string (format "%s-%02d-%02d" year month day)]
    #_(println :first-name first-name :last-name last-name :gender gender-code :date date-string)
    {"gender" {"genderCode" gender-code}
     "familyName" last-name
     "givenName" first-name
     "dateOfBirth" date-string
     "personIdentifiers" {"identifier" (str domain-namespace "_" entity-id)
                          "identifierDomain" {"namespaceIdentifier" domain-namespace}}}))



(defn add-musician [skey data domain-namespace]
  (let [p (musician-to-person data domain-namespace)
        json (json/write-str p)]
    (add-person-json skey json)))

(defn add-musicians
  ([skey db domain-namespace]
   (let [them (two-named-gendered-with-date db)]
     (doseq [t them]
       (println :name (second t))
       (add-musician skey t domain-namespace))))
  ([skey db]
   (add-musicians skey db "Bjond")))
  

        
                      


src/openempi_demo/core.clj
(ns openempi-demo.core
  (:gen-class)
  (:require [zprint.core :as zp]
            [clojure.data.json :as json]
            [clj-http.client :as http]
            [clojure.data.xml :as xml]))

;; --- some utilities

;; Courtesy of : https://nakkaya.com/2010/03/27/pretty-printing-xml-with-clojure/
(defn ppxml
  "Returns a pretty printed xml (string)"
  [xml]
  (let [in (javax.xml.transform.stream.StreamSource.
            (java.io.StringReader. xml))
        writer (java.io.StringWriter.)
        out (javax.xml.transform.stream.StreamResult. writer)
        transformer (.newTransformer 
                     (javax.xml.transform.TransformerFactory/newInstance))]
    (.setOutputProperty transformer 
                        javax.xml.transform.OutputKeys/INDENT "yes")
    (.setOutputProperty transformer 
                        "{http://xml.apache.org/xslt}indent-amount" "2")
    (.setOutputProperty transformer 
                        javax.xml.transform.OutputKeys/METHOD "xml")
    (.transform transformer in out)
    (-> out .getWriter .toString)))

(defn from-json
  "Translates json (string) to Clojure object (edn)"
  [json]
  (json/read-str json :key-fn keyword))

(defn get-body
  "Returns the body of a http response as a clojure data structure"
  [response]
  (-> response :body from-json))

(defn zp-body
  "Pretty prints the body of a http response"
  [response]
  (-> response get-body zp/zprint))

;; --- authentication/session-key

(defn auth-json [username pword]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/security-resource/authenticate"
        json (json/write-str {"username" username
                              "password" pword})]
    #_(println :json json)
    (http/put url {:headers {"Content-type" "application/json"}
                   :body json})))

(defn auth-xml [username pword]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/security-resource/authenticate"
        body (format "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>
<authenticationRequest><username>%s</username><password>%s</password></authenticationRequest>"
                     username
                     pword)]
    (http/put url {:headers {"Content-type" "application/xml"}
                   :body body})))

;; --- person management

(defn add-person-json [session-key json]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/person-manager-resource/addPerson"]
    (http/put url {:headers {"OPENEMPI_SESSION_KEY" session-key
                             "Content-type" "application/json"}
                   :accept :json
                   :body json})))

(defn add-person-xml [session-key xml]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/person-manager-resource/addPerson"]
    (http/put url {:headers {"OPENEMPI_SESSION_KEY" session-key
                             "Content-type" "application/xml"}
                   :accept :xml
                   :body xml})))

;; JSON data should be a "person" as
;; defined in openempi.
;; Some important required fields:
;; 1. dateChanged (must be later than the original dataChanged)
;; 2. personId
;; Returns 204 (No Content) when update is successful.
(defn update-person-json [session-key json]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/person-manager-resource/updatePerson"]
    (http/put url {:headers {"OPENEMPI_SESSION_KEY" session-key
                             "Content-type" "application/json"}
                   :accept :json
                   :body json})))

(defn update-person-xml [session-key xml]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/person-manager-resource/updatePerson"]
    (http/put url {:headers {"OPENEMPI_SESSION_KEY" session-key
                             "Content-type" "application/xml"}
                   :accept :json
                   :body xml})))

;; returns 204 (No Content) ... no matter what... so far.
(defn delete-person [session-key json]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/person-manager-resource/deletePerson"]
    (http/put url {:headers {"OPENEMPI_SESSION_KEY" session-key
                             "Content-type" "application/json"}
                   :accept :json
                   :body json})))

;;; --- person query

(defn find-person-by-id [session-key id]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/person-query-resource/findPersonById"
        data (xml/sexp-as-element ["personIdentifier" {}
                                   ["identifier" {} id]
                                   ["identifierDomain" {}
                                    #_["namespaceIdentifier" {} "Bjond"]
                                    ["universalIdentifier" "2.16.840.1.113883.3.6292.123.7865"]]])
        str (xml/emit-str data)]
    (http/post url {:headers {"OPENEMPI_SESSION_KEY" session-key
                              "Content-type" "application/xml"}
                    :accept :json
                    :body str})
    ))

(defn find-match [session-key xml]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/person-query-resource/findMatchingPersonsByAttributes"]
    (http/post url {:headers {"OPENEMPI_SESSION_KEY" session-key
                              "Content-type" "application/xml"}
                    :accept :json
                    :body xml})))

;; --- some entity API stuff

;; Hard-coding person entity ID to 2 for now.
(defn find-person-record [session-key record-id]
  (let [url (str "http://localhost:8090/openempi-admin/openempi-ws-rest/1.0/records/?entityId=2&recordId="
                 record-id)]
    (http/get url {:headers {"OPENEMPI_SESSION_KEY" session-key}
                   :accept :json})))

(defn entities [session-key]
  (let [url "http://localhost:8090/openempi-admin/openempi-ws-rest/1.0/entities"]
    (http/get url {:headers {"OPENEMPI_SESSION_KEY" session-key}
                   :accept :json})))

;; --- main (unused)

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))


